### PR TITLE
Added instanced tile specs

### PR DESCRIPTION
**Pull request by lilleyse**
For #3177 

Added tests for `Instanced3DModel3DTileContentProvider`.

### CODE REVIEW

**Comment by lilleyse** [comment]
I have a few more tests to add before this is ready.

File: Specs/Data/Tiles3D/instanced/instancedEmptyGltf/tiles.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"properties":{"id":{"minimum":0,"maximum":99},"height":{"minimum":0.7289306027814746,"maximum":97.26047022268176}},"geometricError":40,"root":{"box":[-1.319970476767974,0.6986942109000001,-1.3199295232320258,0.6987257891,0,15],"geometricError":0,"content":{"url":"instancedEmptyGltf.i3dm"}}}
```
Comment (pjcozzi) [inline Specs/Data/Tiles3D/instanced/instancedEmptyGltf/tiles.json]: Can we use human-readable formatting for all tiles.json examples.  There should be a command-line option on our tool.  For the existing content, I'm sure Sublime or Web Storm can do it, or use https://jsonformatter.curiousconcept.com/

File: Specs/Data/Tiles3D/instanced/instancedGltfEmbedded/tiles.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"properties":{"id":{"minimum":0,"maximum":99},"height":{"minimum":0.7289306027814746,"maximum":97.26047022268176}},"geometricError":40,"root":{"box":[-1.319970476767974,0.6986942109000001,-1.3199295232320258,0.6987257891,0,15],"geometricError":0,"content":{"url":"instancedGltfEmbedded.i3dm"}}}
```
Comment (pjcozzi) [inline Specs/Data/Tiles3D/instanced/instancedGltfEmbedded/tiles.json]: For naming, I have been using `Cesium3DTiles` everywhere, which I do not like because it is so verbose.  This uses `Tiles3D`.  Perhaps that is better.  We should be consistent (file renames can be in a separate PR).  What do you suggest?

Comment (lilleyse) [inline Specs/Data/Tiles3D/instanced/instancedGltfEmbedded/tiles.json]: Not sure, but I can rename the folder to `Cesium3DTiles` for now. `Batched3DModel3DTileContentProvider` for example wouldn't be as clear as `Batched3DModelTiles3DContentProvider`. That name itself is too long anyways, and could also change.

**Comment by pjcozzi** [comment]
Cesium is not 100% consistent, but directory names should be Pascal Case.  Can you please update `Data\Tiles3D\instanced` and subdirectories?

File: Specs/Data/Tiles3D/instanced/instancedGltfEmbedded/tiles.json (hunk 2 above)
Comment (pjcozzi) [inline Specs/Data/Tiles3D/instanced/instancedGltfEmbedded/tiles.json]: OK, will revisit.

File: Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+    afterAll(function() {
+        scene.destroyForSpecs();
+    });
+
+    afterEach(function() {
+        scene.primitives.removeAll();
+    });
+
+    function verifyRender(tileset) {
```
Comment (pjcozzi) [inline Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js]: As we add more 3D Tiles tests, we may want to move this and friends to a utility class/function(s) in a separate file(s).

File: Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+    function loadTileset(url) {
+        var tileset = scene.primitives.add(new Cesium3DTileset({
+            url : url
+        }));
+
+        return pollToPromise(function() {
+            // Render scene to progressively load the content
+            scene.renderForSpecs();
+            return tileset.ready && (tileset._root.isReady());
```
Comment (pjcozzi) [inline Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js]: Later, I _think_ we will provide public immutable access to the tree.  I don't know what it will look like yet, so reaching into `_root` is fine now.

File: Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        }).then(function() {
+            return tileset;
+        });
+    }
+
+    function loadTileExpectProcessing(url) {
```
Comment (pjcozzi) [inline Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js]: For future format testing, we can probably make this generic by passing in a function to construct the specific tile format.

File: Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        var tileset = {};
+        var tile = {};
+        var instancedTile = new Instanced3DModel3DTileContentProvider(tileset, tile, url);
+        return loadArrayBuffer(url).then(function(arrayBuffer) {
+            instancedTile.initialize(arrayBuffer);
+            // Expect to stay in the processing state
```
Comment (pjcozzi) [inline Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js]: Copy this comment here:
```
// Expect the tile to never reach the ready state due to returning early in ModelInstanceCollection
```
Otherwise, I could not figure out what this was doing...and why 10.

Comment (lilleyse) [inline Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js]: I might just move the function into the spec itself.

Comment (pjcozzi) [inline Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js]: Yeap, that is probably best.

File: Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        var instancedTile = new Instanced3DModel3DTileContentProvider(tileset, tile, url);
+        return loadArrayBuffer(url).then(function(arrayBuffer) {
+            expect(function() {
+                instancedTile.initialize(arrayBuffer);
+                instancedTile.update(tileset, scene.frameState);
+            }).toThrowDeveloperError();
+        });
+    }
+
+    it('throws with invalid magic', function() {
```
Comment (pjcozzi) [inline Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js]: Instead of having sample data on disk for all the failure cases, is it better to procedurally create the arraybuffer in the test?

**Comment by pjcozzi** [comment]
That's all of my comments.  This is off to a great start.  Let me know when it is ready for a second review.

File: Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js (hunk 7 above)
Comment (lilleyse) [inline Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js]: Yeah this is easy to do for the most part.

**Comment by lilleyse** [comment]
This is ready for a second review. In my next PR I'll make a separate file with common function.

**Comment by pjcozzi** [comment]
This is beautiful.

Coverage for `Instanced3DModel3DTileContentProvider` is only 89%.  Can we test the failure case for the promises, and the `destroy` function?

For the `destroy` function, we could explicitly call it like [here](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Specs/Scene/ModelSpec.js#L481) or implicitly destroy it with the tileset.  For the later, we need to implement `destroy` for `Cesium3DTileset`, which requires some careful tree traversal and taking into account pending requests.  I can help if you want to tackle that now.  Explicitly calling `destroy` is OK too.

**Comment by lilleyse** [comment]
I added some more tests for ready promises and destroy.

To get the `Model` ready promise rejection to work, I modified `Model` so that the promise is rejected in the load error callback instead of in `update`. Otherwise the promise is never rejected. This seems to be the pattern in imagery/terrain providers. 

There's some inconsistency about when `Model` should reject promises and set its state to `FAILED`. Most of the `DeveloperErrors` don't handle that.

I started to look at `Cesium3DTileset.destroy` but decided to hold off until I'm working on the request scheduler. So I kept it simple here.

**Comment by pjcozzi** [comment]
This all looks good.

> There's some inconsistency about when Model should reject promises and set its state to FAILED. Most of the DeveloperErrors don't handle that.

Can you update these cases - including tests - in a separate PR to master and 3d-tiles?
* `throw new RuntimeError('Unsupported glTF Extension: ' + extension);`
* `throw new DeveloperError('bgltf is not a valid Binary glTF file.');`
* `throw new DeveloperError('Only Binary glTF version 1 is supported.  Version ' + version + ' is not.');`

**Comment by lilleyse** [comment]
Will do.
